File: Linked List Cycle/Linked_List_Cycle.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def hasCycle(self, head) -> bool:
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

def create_linked_list(values, pos):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    cycle_node = None

    if pos == 0:
        cycle_node = head

    for i in range(1, len(values)):
        current.next = ListNode(values[i])
        current = current.next

        if i == pos:
            cycle_node = current

    if pos != -1:
        current.next = cycle_node

    return head

def main():
    values = [3, 4, 0, -4]
    pos = 1
    head = create_linked_list(values, pos)
    sol = Solution()
    print(sol.hasCycle(head))

    values2 = [1,2]
    pos2 = 0
    head2 = create_linked_list(values2, pos2)
    print(sol.hasCycle(head2))

    values3 = [1]
    pos3 = -1
    head3 = create_linked_list(values3, pos3)
    print(sol.hasCycle(head3))

if __name__ == "__main__":
    main()

File: Linked List Cycle/Linked_List_Cycle.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }

        return false;
    }
}

public class Main{

    public  static ListNode createLinkedList(int[] values, int pos) {
        if (values.length == 0) {
            return null;
        }

        ListNode head = new ListNode(values[0]);
        ListNode current = head;
        ListNode cycleNode = null;

        if (pos == 0) {
            cycleNode = head;
        }

        for (int i = 1; i < values.length; i++) {
            current.next = new ListNode(values[i]);
            current = current.next;

            if (i == pos) {
                cycleNode = current;
            }
        }

        if (pos != -1) {
            current.next = cycleNode;
        }

        return  head;

    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        int[] values1 = {3, 4, 0, -4};
        int pos1 = 1;
        ListNode head1 = createLinkedList(values1, pos1);
        System.out.println(sol.hasCycle(head1));

        int[] values2 = {1, 2};
        int pos2 = 0;
        ListNode head2 = createLinkedList(values2, pos2);
        System.out.println(sol.hasCycle(head2));

        int[] values3 = {1};
        int pos3 = -1;
        ListNode head3 = createLinkedList(values3, pos3);
        System.out.println(sol.hasCycle(head3));

    }

}
